File: python/say/say.py
def say(number):
    if number < 0 or number > 999_999_999_999:
        raise ValueError(r".+")

    return say_billion(number)


def say_x(number, x, text, f):
    if number < x:
        return f(number)
    elif number % x == 0:
        return f(number // x) + " " + text
    else:
        return f(number // x) + " " + text + " " + f(number % x)


def say_hundred(number):
    return say_x(number, 100, "hundred", say_0_99)


def say_thousand(number):
    return say_x(number, 1_000, "thousand", say_hundred)


def say_million(number):
    return say_x(number, 1_000_000, "million", say_thousand)


def say_billion(number):
    return say_x(number, 1_000_000_000, "billion", say_million)


def say_0_99(number):
    nums_le20 = ["zero", "one", "two", "three", "four",
                 "five", "six", "seven", "eight", "nine",
                 "ten", "eleven", "twelve", "thirteen", "fourteen",
                 "fifteen", "sixteen", "seventeen", "eighteen", "nineteen", ]
    tens = ["twenty", "thirty", "forty", "fifty",
            "sixty", "seventy", "eighty", "ninety"]

    if number < 20:
        return nums_le20[number]
    elif number % 10 == 0:
        return tens[number // 10 - 2]
    else:
        return say_0_99(number // 10 * 10) + "-" + say_0_99(number % 10)

File: haskell/say/src/Say.hs
module Say
  ( inEnglish
  ) where

inEnglish :: Integer -> Maybe String
inEnglish n | n < 0     = Nothing
            | otherwise = Just (sayBillion n)
 where
  sayBillion  = say sayMillion "billion" 1000000000
  sayMillion  = say sayThousand "million" 1000000
  sayThousand = say sayHundred "thousand" 1000
  sayHundred  = say say99 "hundred" 100

say :: (Integer -> String) -> String -> Integer -> Integer -> String
say f text x n
  | n < x          = f n
  | n `mod` x == 0 = f (n `div` x) ++ " " ++ text
  | otherwise      = f (n `div` x) ++ " " ++ text ++ " " ++ f (n `mod` x)

say99 :: Integer -> String
say99 0  = "zero"
say99 1  = "one"
say99 2  = "two"
say99 3  = "three"
say99 4  = "four"
say99 5  = "five"
say99 6  = "six"
say99 7  = "seven"
say99 8  = "eight"
say99 9  = "nine"
say99 10 = "ten"
say99 11 = "eleven"
say99 12 = "twelve"
say99 13 = "thirteen"
say99 15 = "fifteen"
say99 18 = "eighteen"
say99 20 = "twenty"
say99 30 = "thirty"
say99 40 = "forty"
say99 50 = "fifty"
say99 80 = "eighty"
say99 n | n `elem` [14, 16, 17, 19] = say99 (n `mod` 10) ++ "teen"
        | n `elem` [60, 70, 90] = say99 (n `div` 10) ++ "ty"
        | otherwise = say99 (n `div` 10 * 10) ++ "-" ++ say99 (n `mod` 10)
